pkg/yum/yum.go

```diff
@@ -44,32 +56,157 @@ type Location struct {
 	Href string `xml:"href,attr"`
 }
 
-// ExtractPackageData returns array of package information given URL to a yum repo.
-func ExtractPackageData(client http.Client, url string) ([]Package, error) {
-	var repomd Repomd
-	var primaryURL string
+type Repository struct {
+	Client         http.Client   // Client to fetch repository data
+	URL            string        // URL of the repository
+	Packages       []Package     // Packages repository contains
+	GPGKey         *string       // GPG key of the repository
+	Signature      *string       // Signature of the repository
+	Repomd         Repomd        // Repomd of the repository
+	RequestTimeout time.Duration // Maximum time to wait for response while fetching
+}
+
+// NewRepository returns a new Repository whose URL is url
+func NewRepository(client http.Client, url string, timeout time.Duration) Repository {
+	return Repository{
+		URL:            url,
+		Packages:       nil,
+		Repomd:         Repomd{},
+		Client:         client,
+		RequestTimeout: timeout,
+	}
+}
+
+// FetchRepomd populates r.Repomd with repository's repomd.xml metadata. Returns response code and error.
+func (r *Repository) FetchRepomd() (int, error) {
+	var result Repomd
 	var err error
+	var resp *http.Response
+	var repomdURL string
 
-	if repomd, err = GetRepomdXML(client, url); err != nil {
-		return []Package{}, fmt.Errorf("Error parsing repomd.xml: %v", err)
+	if repomdURL, err = r.getRepomdURL(); err != nil {
+		return 0, fmt.Errorf("GET error: %w", err)
+	}
+	if resp, err = fetchWithTimeout(repomdURL, r.RequestTimeout); err != nil {
+		return 0, fmt.Errorf("GET error: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return resp.StatusCode, fmt.Errorf("Status error: %v", resp.StatusCode)
+	}
+	if result, err = parseRepomdXML(resp.Body); err != nil {
+		resp.Body.Close()
+		return resp.StatusCode, fmt.Errorf("error parsing repomd.xml: %w", err)
 	}
 
-	if primaryURL, err = GetPrimaryURL(repomd, url); err != nil {
-		return []Package{}, fmt.Errorf("Error getting primary URL: %v", err)
+	r.Repomd = result
+	return resp.StatusCode, err
+}
+
+// FetchPackages populates r.Packages with metadata of each package in repository. Returns response code and error.
+func (r *Repository) FetchPackages() (int, error) {
```

Comment:
You might consider renaming these methods from  FetchXYZ() to just  XYZ()

Reply:
I can, but I named it this way to imply that this method makes a request to a URL